src/platform/nxp/common/ConfigurationManagerImpl.cpp

```diff
@@ -0,0 +1,316 @@
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *          Provides the implementation of the Device Layer ConfigurationManager object
+ *          for NXP platforms using the NXP SDK.
+ */
+
+/* this file behaves like a config.h, comes first */
+#include <platform/internal/CHIPDeviceLayerInternal.h>
+#include <app/server/Server.h>
+#include <platform/ConfigurationManager.h>
+#include "NXPConfig.h"
+#include <platform/internal/GenericConfigurationManagerImpl.ipp>
+#include <platform/DiagnosticDataProvider.h>
+
+#include "fsl_device_registers.h"
+
+#if CONFIG_CHIP_PLAT_LOAD_REAL_FACTORY_DATA
+#include "FactoryDataProvider.h"
+#endif
+
+#if CHIP_DEVICE_CONFIG_ENABLE_WPA
+extern "C" {
+#include "wlan.h"
+}
+#endif
+
+namespace chip {
+namespace DeviceLayer {
+
+using namespace ::chip::DeviceLayer::Internal;
+
+ConfigurationManagerImpl & ConfigurationManagerImpl::GetDefaultInstance()
+{
+    static ConfigurationManagerImpl sInstance;
+    return sInstance;
+}
+
+CHIP_ERROR ConfigurationManagerImpl::Init()
+{
+    CHIP_ERROR err;
+    uint32_t rebootCount = 0;
+    bool failSafeArmed;
+
+    // Initialize the generic implementation base class.
+    err = Internal::GenericConfigurationManagerImpl<NXPConfig>::Init();
+    SuccessOrExit(err);
+
+    if (NXPConfig::ConfigValueExists(NXPConfig::kCounterKey_RebootCount))
+    {
+        err = GetRebootCount(rebootCount);
+        SuccessOrExit(err);
+
+        err = StoreRebootCount(rebootCount + 1);
+        SuccessOrExit(err);
+    }
+    else
+    {
+        // The first boot after factory reset of the Node.
+        err = StoreRebootCount(1);
+        SuccessOrExit(err);
+    }
+
+    if (!NXPConfig::ConfigValueExists(NXPConfig::kCounterKey_TotalOperationalHours))
+    {
+        err = StoreTotalOperationalHours(0);
+        SuccessOrExit(err);
+    }
+
+    if (!NXPConfig::ConfigValueExists(NXPConfig::kCounterKey_BootReason))
+    {
+        err = StoreBootReason(to_underlying(BootReasonType::kUnspecified));
+        SuccessOrExit(err);
+    }
+
+    // TODO: Initialize the global GroupKeyStore object here
+
+    // If the fail-safe was armed when the device last shutdown, initiate a factory reset.
+    if (GetFailSafeArmed(failSafeArmed) == CHIP_NO_ERROR && failSafeArmed)
+    {
+        ChipLogProgress(DeviceLayer, "Detected fail-safe armed on reboot; initiating factory reset");
+        InitiateFactoryReset();
```

Comment:
This seems very broken.  If I commission the device into fabric 1 and then open a commissioning window and someone connects over PASE and then partway through that second commissioning the device reboots.... why should that factory reset?